File: day4/program2.hs

{-# OPTIONS_GHC -Wno-incomplete-patterns #-}

{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
import Data.List (intersect, union)

split :: Char -> String -> [String]
split _ "" = []
split delimiter str =
    let (start, rest) = break (== delimiter) str
        (_, remain) = span (== delimiter) rest
     in start : split delimiter remain


first (a:b) = a
second (a:b:_) = b

lineRanges line =
    let ranges = split ',' line
        r1 = first ranges
        r2 = second ranges
        rangeOneTemp = split '-' r1
        rangeTwoTemp = split '-' r2
        r1p1 = first rangeOneTemp
        r1p2 = second rangeOneTemp
        r2p1 = first rangeTwoTemp
        r2p2 = second rangeTwoTemp
        rangeOne = [(read r1p1::Int)..(read r1p2::Int)]
        rangeTwo = [(read r2p1::Int)..(read r2p2::Int)]
        rangeOneLength = length rangeOne
        rangeTwoLength = length rangeTwo
        maxLen = max rangeOneLength rangeTwoLength
        theUnion = intersect rangeOne rangeTwo
    in
        not (null theUnion)


calc :: Foldable t => t [Char] -> Int
calc = foldl
    ( \total x ->
        if lineRanges x then
            total + 1
        else
            total
    )
    0

main :: IO ()
main = do
  fileData <- readFile "input.txt"
  let strs = lines fileData
  let sum = calc strs
  print sum

File: day8/program2.py
lines = []
with open("input.txt") as file:
    for line in file:
        lines.append(line.rstrip())

grid =  []

for line in lines:
    row = []
    parts = list(line)
    for tree in parts:
        row.append(tree)
    grid.append(row)

def score(input, x, y, dx, dy):
    h = input[x][y]
    x += dx
    y += dy
    sum = 0
    while (y >= 0 and y < len(input) and x >= 0 and x < len(input[y])):
        sum +=1
        if (input[x][y] >= h):
            return sum
        x += dx
        y += dy

    return sum

sum = 0

for i in range(0, len(lines)):
    for j in range(0, len(lines)):
        v1 = score(grid, i,j, 1, 0)
        v2 = score(grid, i,j, 0, 1)
        v3 = score(grid, i,j, -1, 0)
        v4 = score(grid, i,j, 0,-1)
        total = v1*v2*v3*v4
        sum = max(total, sum)

print(sum)
